MovieApp/AppViewController.swift

```diff
@@ -193,141 +276,115 @@ class AppViewController: UIViewController {
-        job_tag.leadingAnchor.constraint(equalTo: view1.leadingAnchor).isActive = true
-        job_tag.trailingAnchor.constraint(equalTo: view1.trailingAnchor).isActive = true
-        job_tag.bottomAnchor.constraint(equalTo: view1.bottomAnchor).isActive = true
-        
-        return view1
-    }
-    
-    
-    private func updateContentSizeScrollView() {
-        DispatchQueue.main.async {
-            self.scrollView.contentSize.height = self.mainView.frame.size.height
+extension AppViewController : NetworkServiceProtocol {
+    func getDataFromURL(requestUrl: String) {
+        guard let url = URL(string: requestUrl) else { return  }
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        let networkService = NetworkService()
+        networkService.executeUrlRequest(request) { dataValue, error in
+            if error != nil {
+                self.buildErrorView()
+                return
+            }
+            guard let value = dataValue else {
+                return
+            }
+            
+            guard let value = try? JSONDecoder().decode(TMDBMovieModel.self, from: value) else {
+                self.buildErrorView()
+                return
+            }
+            
+            let url = URL(string: "https://image.tmdb.org/t/p/original"+value.posterPath)
+            DispatchQueue.global().async {
+                if let data = try? Data(contentsOf: url!) {
+                    DispatchQueue.main.async {
+                        self.imageView.image = UIImage(data: data)!
+                    }
+                }
+            }
+            
+            self.rating_label.text = String(Int(value.voteAverage*10))+"%"
+            self.movie_title_label.text = value.title
+            self.movie_release_date_label.text = value.releaseDate
+            var movieTags = ""
+            var counter = 0
+            for genre in value.genres {
+                if counter > 3 {
+                    break
+                }
+                movieTags += genre.name + "  "
+                counter += 1
+            }
+            self.movie_tags_label.text = movieTags
+            self.movie_duration_label.text = String(value.runtime) + " m"
+            self.overview_text.text = value.overview
+            self.castCollectionView.getDataFromURL(requestUrl: "https://api.themoviedb.org/3/movie/"+String(value.id)+"/credits?api_key="+self.apiKey)
         }
     }
-    
-    
-    private func addBackgroundImage() {
-        let image = UIImage(named: "ironman_cover")
-        textOverImageView.backgroundColor = UIColor(patternImage: image!)
-        
-    }
+}
+
 
+extension AppViewController : appViewCastCollectionCommunication {
+    func castDataLoaded() {
+        DispatchQueue.main.async {   // Hacky solution, please inform me if there is a better solution
```

Comment:
`appViewCastCollectionCommunication` is a strange name, not suggesting what this protocol really does. I would say that it suggest that the protocol is not needed. You want to send the information to the view controller that the collection view has finished loading data. The problem here is that the collection view is not the one who should load the data, as I explained in the previous comment. 

The view controller is the one calling the network service for data, reloading the collection view and then the view controller can, if needed, call the 
```
self.castCollectionView.snp.makeConstraints{ (make) in
   make.height.equalTo(self.castCollectionView.collectionView.contentSize.height)
}
 ```
block of code, if that is needed. This block of code is not that hacky as is this protocol usage. But removing the collectionView altogether here and using only Stack view would probably remove the need for this block of code too.